### DISCUSSION TITLE
Drop column queries are not generated

### DISCUSSION THREAD

**Discussion by NLin4com** [mikro-orm/mikro-orm#4660] [General]
**Describe the bug**
When I add a new column in my schema and generate migration, only `up` function is generated, but `down` one is not. 

When I remove a column from the schema and try to generate migration only with these changes, I get a message that no changes required. But if I add another change that can be handled by micro-orm, add column request is generated in `down` function for column that was removed from schema before

**To Reproduce**
1. Add a new column to schema
2. Generate migration

Or
1. Remove a column from schema
2. Try to generate migration

**Expected behavior**
Both `up` and `down` functions have to be generated in case of adding/dropping columns. When one adds a column, `up` function has to contain add column query and `down` function has to contain drop column query. When one removes a column from schema - vice versa

**Additional context**
I use plain JS approach, not Typescript, my schemas are described separately from my entities but in the same file.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.3.0 |
| mikro-orm | 5.7.14 |
| mikro-orm/postgresql | 5.7.14 |

**Comment by B4nan** [comment]
I am not able to reproduce this. Please provide at least entity definitions and your ORM config.

FYI schema diffing is well-tested, you can see the current tests here:

https://github.com/mikro-orm/mikro-orm/tree/master/tests/features/schema-generator

One test that does pretty much what you say is not working is here:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/schema-generator/adding-composite-fk.postgres.test.ts

In the snapshot you can see the up and down queries do have the correct shape.

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/schema-generator/__snapshots__/adding-composite-fk.postgres.test.ts.snap#L34-L46

**Reply by NLin4com** [reply]
Here's an example of one of my entities:

```
const { EntitySchema } = require("@mikro-orm/core");

class DictionaryAudioCodec {
  id;
  codec;
  isDefault;
  isDeleted;

  constructor(
    id,
    codec,
    isDefault,
    isDeleted
  ) {
    this.id = id;
    this.codec = codec;
    this.isDefault = isDefault;
    this.isDeleted = isDeleted;
  }
}

module.exports = DictionaryAudioCodec
module.exports.schema = new EntitySchema({
  class: DictionaryAudioCodec,
  tableName: "av_dic_audio_codec",
  properties: {
    id: {
      primary: true,
      type: "integer",
      autoincrement: true
    },
    codec: {
      type: "string",
      length: 11,
      nullable: false
    },
    isDeleted: {
      type: "boolean",
      default: false,
      nullable: false
    },
    isDefault: {
      type: "boolean",
      default: false,
      nullable: false
    }
  }
});
```

If I add a column `isDefault2` like this:

```
isDefault2: {
      type: "boolean",
      default: false,
      nullable: false
    }
```

The following migration is generated:

```
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
const { Migration } = require('@mikro-orm/migrations');

class Migration20230828183741 extends Migration {

  async up() {
    this.addSql('alter table "av_dic_audio_codec" add column "is_default2" boolean not null default false;');
  }

}
exports.Migration20230828183741 = Migration20230828183741;
```

Maybe something is wrong with my schema definition?

**Reply by B4nan** [reply]
The schema looks good to me. How do you create the migration? How does your ORM config look like?

**Reply by NLin4com** [reply]
@B4nan problem was with my config file. I set flag `safe` to `true`. That's why my migrations didn't contain drop requests. As soon as I set it to false, script start adding drop requests. Thank you for your help!

**Reply by davidzwa** [reply]
Hmm Im never getting any drop methods, nor are columns dropped. 
Im using SWC together with a NodeJS project in CommonJS mode. Database better-sqlite3. MikroORM 5.8.1 (better-sqlite, core, cli and migrations packages).
My safe property is set to false.

Im prepping a repro.

**Reply by B4nan** [reply]
That's pretty weird, as that method literally calls `dropSchema` first before creating it again:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/AbstractSchemaGenerator.ts#L42

edit: oops, confused two issues, I though this is another one that used `refreshDatabase` explicitly.

**Reply by davidzwa** [reply]
Instead of a repro, I thought it easier to debug the create migration javascript.

I can confirm the `safe: false` prop is propagated, as well as the `dropTables: true` setting.

Somehow the migration up-down diff is empty when I remove a column from a registered entity (Im also definitely running in ts-node, so its not a transpilation inconsistency). 
Also when I create a migration the down prop is always empty (and no down method is therefore created).

![image](https://github.com/mikro-orm/mikro-orm/assets/6005355/72683e23-2026-40bf-88dd-07fa6fdd78a2)


Ill report later. Once I've drilled into the SchemaGenerator to see what is going wrong.

**Reply by B4nan** [reply]
Down migrations are not supported for SQLite, as it has _very_ limited abilities, you literally need to recreate a table to be able to drop a column from it.

**Reply by davidzwa** [reply]
Ah wow ok that is crucial information. Thanks! 🎖️

**Reply by B4nan** [reply]
This is the switch that drives it, feel free to try to change the value to true locally to see how it works, current versions of knex have some kind of support for this, but unfortunately it wasn't suitable for the ORM. I'd like to lift this restriction at some point.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/sqlite/src/SqlitePlatform.ts#L116

**Reply by davidzwa** [reply]
I am assuming you're aware of the DROP COLUMN availability beyond SQLite version 3.35?
https://www.quickprogrammingtips.com/sqlite/how-to-add-or-remove-columns-from-sqlite-tables.html

Thanks for the pointer btw!

[EDIT] the flag indeed does enable the down method, still the schema up diff is empty on column removal. I think I should be looking at Knex for that.

[EDIT2] ah I see the SchemaGenerator is still the ORMs territory.

**Reply by B4nan** [reply]
IIRC it's not about that, it's about the restrictions sqlite brings on dropping the columns, e.g. to be able to drop a column, it had to be nullable, similarly adding a column to existing table had to be nullable, and there was more. But it's been years since I reviewed how this works, so maybe it's no longer true, need to give it a fresh try I guess.

The bottom line is, that the schema diffing is built on top of knex and as such, knex would first have to support this out of box. And since there wasn't much movement in knex in last months, I don't expect that to happen any time soon. I've been thinking about getting rid of knex for some time, right now I have better things to focus on, but maybe next major will just do that.

**Reply by davidzwa** [reply]
![image](https://github.com/mikro-orm/mikro-orm/assets/6005355/baee66f3-f696-46a2-9d45-c0a4ed66215a)
Wow ok I think I get what you mean. This was a nullability change migration to get the database into the nullable shape for testing your statement.

**Reply by B4nan** [reply]
Yeah, but that's how knex is handling things because of those restrictions, that itself is not telling us if it's still needed or not. Can you try to run the queries manually to check? Add a non null column to existing table, then change its nullability back and forth, then try to drop it. (and I would try this on a table with some data ideally)

I would still guess those restrictions are there (although I haven't seen them mentioned in the sqlite docs nowadays, but I haven't been trying that hard to find it), as the knex support for this is relatively new, definitely newer than that sqlite version you mentioned.

**Reply by davidzwa** [reply]
(I did this before reading your response). 

After applying the nullability change (non-nullable => nullable int column), and running migration create, I tried to drop the now-nullable column to force a drop statement, but to no avail. No up-diff is generated and consequently the output is `No changes required, schema is up-to-date`. 

FYI snapshot: true
FYI#2 platform down flag is true (most likely unrelated)

**Reply by davidzwa** [reply]
Manually applying a DROP COLUMN works fine. I can live with that for now.

**Reply by B4nan** [reply]
FYI as part of #4884 I've enabled the drop column queries in sqlite, had to do it explicitly as it is apparently skipped on knex level. 

(down migrations are still not enabled for other reasons, I hope I can lift that restriction too at some point)
